### PR TITLE
zephyr: ll-schedule: fix a race when unregistering the domain

### PR DESCRIPTION
**Pull request by lyakh**
The following race has been observed:
```
core 0				core 3

				timer IRQ
				spin_lock(&timeout_lock)
IPC IRQ
task remove
zephyr_domain_unregister()
spin_lock(&timeout_lock)
				timer remove
				spin_unlock(&timeout_lock)
k_timer_stop()
z_abort_timeout()
				z_timer_expiration_handler()
				z_add_timeout()
```
which leads to the LL domain timer being added needlessly and to a later lock up and an IPC timeout on the host. Fix this by removing the timer in such cases.

BugLink: https://github.com/thesofproject/sof/issues/6173

### CODE REVIEW

**Review by paulstelian97 — APPROVED** [review]
Feels harmless and like a good fix. Still worth checking if nothing important is broken by this (shouldn't happen but better safe than sorry).

I'll assume no such issues are going to pop up though, so approving.

**Comment by lgirdwood** [comment]
@marcinszkudlinski @mwasko pls check and test. This improves stability a lot with the reproduction stress test.

File: src/schedule/zephyr_domain.c [hunk 1]
```diff
@@ -100,8 +101,18 @@ static void zephyr_domain_timer_fn(struct k_timer *timer)
 	struct zephyr_domain *zephyr_domain = k_timer_user_data_get(timer);
 	int core;
 
-	if (!zephyr_domain)
+	/*
+	 * A race is possible when the Zephyr LL scheduling domain is being
+	 * unregistered while a timer IRQ is processed on a different core. Then
+	 * the timer is removed by the former but then re-added by the latter,
+	 * but this time with no user data and no handler set. This leads to the
+	 * timer continuint to trigger and then leading to a lock up when it is
```
Comment (kv2019i) [inline src/schedule/zephyr_domain.c:109]: typoe: continuint

**Review by kv2019i — APPROVED** [review]
Good find @lyakh !

**Comment by lyakh** [comment]
@wszypelt what happened to the TLG IPC4 tests? Otherwise CI looks good. The single on-device failure is for verify-kernel-boot-log.sh

**Comment by wszypelt** [comment]
@lyakh something broke in building, but it looks like now everything is ok

File: src/schedule/zephyr_domain.c (hunk 1 above)
Comment (lyakh) [reply src/schedule/zephyr_domain.c]: fixed

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@lrudyX @wszypelt CI has been pending a long time. Are we good to merge ?

**Pull request merged**